### ISSUE TITLE
NullRef in Darc's DisambiguateAssetsAsync

### ISSUE THREAD

**Issue by crummel** [dotnet/arcade#6651]
While trying to use `darc update-dependencies --id 71304` to update source-build's Version.Details.xml, @dseefeld and I ran into this exception:

```
fail: Microsoft.DotNet.Darc.Operations.Operation[0]
      Error: Failed to update dependencies.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.DotNet.DarcLib.Remote.DisambiguateAssetsAsync(IRemoteFactory remoteFactory, Dictionary`2 buildCache, Dictionary`2 nugetConfigCache, DependencyDetail parentCoherentDependency, DependencyDetail cpdDependency) in C:\temp\dotnet_arcade-services\src\Microsoft.DotNet.Darc\src\DarcLib\Actions\Remote.cs:line 705
   at Microsoft.DotNet.DarcLib.Remote.GetRequiredStrictCoherencyUpdatesAsync(IEnumerable`1 dependencies, IRemoteFactory remoteFactory) in C:\temp\dotnet_arcade-services\src\Microsoft.DotNet.Darc\src\DarcLib\Actions\Remote.cs:line 600
   at Microsoft.DotNet.DarcLib.Remote.GetRequiredCoherencyUpdatesAsync(IEnumerable`1 dependencies, IRemoteFactory remoteFactory, CoherencyMode coherencyMode) in C:\temp\dotnet_arcade-services\src\Microsoft.DotNet.Darc\src\DarcLib\Actions\Remote.cs:line 1057
   at Microsoft.DotNet.Darc.Operations.UpdateDependenciesOperation.CoherencyUpdatesAsync(IRemote barOnlyRemote, IRemoteFactory remoteFactory, List`1 currentDependencies, List`1 dependenciesToUpdate) in C:\temp\dotnet_arcade-services\src\Microsoft.DotNet.Darc\src\Darc\Operations\UpdateDependenciesOperation.cs:line 316
   at Microsoft.DotNet.Darc.Operations.UpdateDependenciesOperation.ExecuteAsync() in C:\temp\dotnet_arcade-services\src\Microsoft.DotNet.Darc\src\Darc\Operations\UpdateDependenciesOperation.cs:line 126
```

I've manually checked the URL generated in [AzureDevOpsClient](https://github.com/dotnet/arcade-services/blob/c93f48d36fb8ab1153bedd56d843e260a72bea03/src/Microsoft.DotNet.Darc/src/DarcLib.AzDev/AzureDevOpsClient.cs#L119) and got a 401 so I'm not sure if the different handling of 400 and 404 has something to do with it.  Fiddler debugging hasn't helped a lot since it seems to interfere with authentication itself.

Let me know if there's any other info that could help troubleshoot this.

**Comment by dleeapho** [comment]
This issue is blocking source-build December servicing for 3.1.11.

**Comment by missymessa** [comment]
Able to get a repro, but it's not quite clear why the gitClient/fileManager don't exist for this scenario. It's blowing up when trying to decide between two builds of templating to use because fileManager object is null.

**Comment by missymessa** [comment]
The `update-dependencies` call implements a BarOnlyRemote, which I believe doesn't expect to do any sort of interactions with GitHub/AzDO. However, for this build, it has to because it finds more than one build to use from Templating.

**Comment by missymessa** [comment]
Okay, chatted with @mmitche about this scenario and he's come up with this solution: 

The failure is occurring at this point in the code: https://github.com/dotnet/arcade-services/blob/master/src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Remote.cs#L705

- [ ] We should add and implement the method for ReadNugetConfigAsync to the IRemote interface and any concrete classes that implement it. (Currently, that method only exists on GitFileManager). After that has been implemented, we will need to refactor the above code to look something like this: 

```
IRemote remoteClient = await remoteFactory.GetRemoteAsync(parentCoherentDependency.RepoUri, _logger);
XmlDocument nugetConfig = await remoteClient.ReadNugetConfigAsync(parentCoherentDependency.RepoUri, parentCoherentDependency.Commit);
```

- [ ] Verify the fix locally and unblock source-build: 
- Clone source-build's repo (check with @dseefeld to see which branch or PR would be the appropriate place to make this change to and get that branch). 
- To verify this fix, set up the debug configuration of the Microsoft.DotNet.Darc project with the following values: 
    - Application arguments: `update-dependencies --id 71304`
    - Working directory: your location of the source-build repo on your local machine
- Run the project in debug mode, verify that the fix is correct and that the dependencies are updated successfully in source-build.
- Work with @dseefeld to get the changes merged into a PR.

**Comment by epananth** [comment]
here is the PR -> https://github.com/dotnet/arcade-services/pull/1490 need to get this approved and merged.

**Comment by sunandabalu** [comment]
This is rolled out on 12/16/2020
